rules/cache/list.go

```diff
@@ -0,0 +1,152 @@
+// Copyright (c) 2017 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package cache
+
+import (
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/m3db/m3collector/id"
+	"github.com/m3db/m3metrics/rules"
+)
+
+// element is a list element
+type element struct {
+	namespace string
+	idHash    id.Hash
+	result    rules.MatchResult
+	deleted   bool
+	expiryNs  int64
+	prev      *element
+	next      *element
+}
+
+func (e *element) ShouldExpire(now time.Time) bool {
+	return time.Unix(0, atomic.LoadInt64(&e.expiryNs)).Before(now)
+}
+
+func (e *element) SetExpiry(t time.Time) {
+	atomic.StoreInt64(&e.expiryNs, t.UnixNano())
+}
+
+// list is a type-specific implementation of doubly linked lists consisting
+// of elements so when we retrieve match results, there is no conversion
+// between interface{} and the concrete result type.
```

Comment:
Yeah possibly. I had the same thoughts about the pools in m3x and talked to @kobolog about it a while back, guess that was put on the back burner atm. The general thoughts was having a template/ package that's used to generate type-specific types and have the go generate commands integrated w/ Makefile. There might be some minor issues with import paths etc, but overall felt doable. Could revisit this if handwriting basic data structures becomes a common occurrence.